### ISSUE TITLE
R_LIBS_USER/R_LIBS_SITE: Add %b conversion specifiers for R_BIOC_VERSION

### ISSUE THREAD

**Issue by HenrikBengtsson** [HenrikBengtsson/Wishlist-for-R#123]
# Wish

In addition to the ones already supported (see `help("libPaths", package = "base")`), add a `%b` conversion specifiers for `R_LIBS_USER` and `R_LIBS_SITE`:

* `%b`: The Bioconductor version per environment variable `R_BIOC_VERSION`
  
This would make it possible to specify version invariant library paths, e.g.

```
R_USER_LIBS=~/R/%p-library/%v-bioc_%b
```

which will, for instance, expand to different paths depending on R version and `R_BIOC_VERSION`, which is set internally in R but can be overridden.  This would allow users to maintain multiple package libraries in parallel, one for each combination of (R, Bioc) version, without much extra work, e.g.

```sh
~/R/x86_64-pc-linux-gnu-library/4.0-bioc_3.12
~/R/x86_64-pc-linux-gnu-library/4.0-bioc_3.13
~/R/x86_64-pc-linux-gnu-library/4.1-bioc_3.13
```

**Comment by HenrikBengtsson** [comment]
# Patch

A patch for this is:

```diff
Index: src/library/base/R/library.R
===================================================================
--- src/library/base/R/library.R	(revision 80146)
+++ src/library/base/R/library.R	(working copy)
@@ -1002,6 +1002,8 @@
     x <- expand(x, "a", R.version$arch)
     ## %o => os
     x <- expand(x, "o", R.version$os)
+    ## %b => R_BIOC_VERSION
+    x <- expand(x, "b", Sys.getenv("R_BIOC_VERSION", NA_character_))
 
     gsub("%%", "%", x, fixed=TRUE)
 }
```

**Comment by HenrikBengtsson** [comment]
Ah... the default value of `R_BIOC_VERSION` is set in [`tools:::.BioC_version_associated_with_R_version()`](https://github.com/wch/r-source/blob/5767e86ab682823c78e17e1a33cc322a7e2a874d/src/library/tools/R/utils.R#L516-L517), so it has to be something like:

```diff
$ svn diff src/library/base/R/library.R
Index: src/library/base/R/library.R
===================================================================
--- src/library/base/R/library.R	(revision 80146)
+++ src/library/base/R/library.R	(working copy)
@@ -1002,6 +1002,8 @@
     x <- expand(x, "a", R.version$arch)
     ## %o => os
     x <- expand(x, "o", R.version$os)
+    ## %b => R_BIOC_VERSION
+    x <- expand(x, "b", tools:::.BioC_version_associated_with_R_version())
 
     gsub("%%", "%", x, fixed=TRUE)
 }
```

However, since this is run at startup, that `.BioC_version_associated_with_R_version()` function would need to be in the **base** package.

So a little bit more work to get this working.

**Comment by rsettlage** [comment]
It would be nice if the OS info included distribution.  This is more important with the use of containers where the libraries could be stored on the host and multiple containers could be in use with different container OS distros and or version, Centos 6/7 or Ubuntu 18 or...

**Comment by HenrikBengtsson** [comment]
Yes, that's an important one too.  There's probably room for lots of more conversion specifiers.  I might be wrong, but I think it also matters whether R was configured and built with `--enable-R-shlib` (shared/dynamic library 'libR') or not.

**Comment by gaborcsardi** [comment]
@rsettlage Is this not good?

```r
> osVersion
[1] "Ubuntu 20.04.2 LTS"
```

**Comment by rsettlage** [comment]
That is perfect, minus the spaces.

From: Gábor Csárdi ***@***.***>
Reply-To: HenrikBengtsson/Wishlist-for-R ***@***.***>
Date: Thursday, April 8, 2021 at 3:36 AM
To: HenrikBengtsson/Wishlist-for-R ***@***.***>
Cc: Bob ***@***.***>, Mention ***@***.***>
Subject: Re: [HenrikBengtsson/Wishlist-for-R] R_LIBS_USER/R_LIBS_SITE: Add %b conversion specifier for R_BIOC_VERSION (#123)


@rsettlage<https://github.com/rsettlage> Is this not good?

> osVersion

[1] "Ubuntu 20.04.2 LTS"

**Comment by HenrikBengtsson** [comment]
@gaborcsardi, I'm not sure I'm following.  The supported conversion specifiers (`?libPaths`) are, e.g.

```r
> cat(base:::.expand_R_libs_env_var("%%V=%V\n%%v=%v\n%%p=%p\n%%a=%a\n%%o=%o\n"))
%V=4.0.4
%v=4.0
%p=x86_64-pc-linux-gnu
%a=x86_64
%o=linux-gnu
```

That's what we get to play with in `R_LIBS_USER` and `R_LIBS_SITE`.
